### PR TITLE
refactor(rpc): migrate RPC to ts

### LINKED ISSUE #1500: Migrate RPC to ts

### PR DESCRIPTION
**Pull request by martinkaintas**
closes #1500 

This PR migrates the RPC util to TS

### CODE REVIEW

File: docs/guides/build-wallet.md [hunk 1]
```diff
@@ -112,7 +112,7 @@ async function init () {
   }).then(wallet => {
     chrome.runtime.onConnect.addListener(async function (port) {
       // create connection
-      const connection = await BrowserRuntimeConnection({ connectionInfo: { id: port.sender.frameId }, port })
+      const connection = await new BrowserRuntimeConnection({ connectionInfo: { id: port.sender.frameId }, port })
```
Comment (davidyuk) [inline docs/guides/build-wallet.md:115]:
```suggestion
      const connection = new BrowserRuntimeConnection({ connectionInfo: { id: port.sender.frameId }, port })
```

File: src/utils/aepp-wallet-communication/connection/browser-runtime.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  }) {
+    if (getBrowserAPI().runtime == null) throw new RpcConnectionError('Runtime is not accessible in your environment')
+    this.debug = debug
+    this.connectionInfo = connectionInfo
+    this.port = port ?? getBrowserAPI().runtime.connect(...[connectionInfo.id ?? undefined])
+  }
+
+  /**
+   * Disconnect
+   * @instance
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/connection/browser-runtime.ts:89]:
```suggestion
```
this shouldn't be needed

File: src/utils/aepp-wallet-communication/connection/index.ts [hunk 3]
```diff
@@ -23,63 +23,44 @@
...
- */
+export default interface WalletConnection{
+  /**
+   * Connect
+   * @instance
+   * @abstract
+   * @param onMessage - Message handler
+   * @param onDisconnect - trigger when runtime connection in closed
+   */
+  connect: (onMessage: Function, onDisconnect?: Function) => void
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/connection/index.ts:44]: `Function` is too general, proposing to specify amount of arguments and `: void`?

File: src/utils/aepp-wallet-communication/connection/index.ts [hunk 4]
```diff
@@ -23,63 +23,44 @@
...
- * @instance
- * @abstract
- * @rtype () => Boolean
- * @return {Boolean} Is connected
- */
+export default interface WalletConnection{
+  /**
+   * Connect
+   * @instance
+   * @abstract
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/connection/index.ts:40]: This is already inside interface, not sure that `abstract` makes sense here

File: src/typings/window/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+
+// ? SEE https://robkendal.co.uk/blog/how-to-fix-property-does-not-exist-on-window-type-in-typescript
+export {}
+
+declare global {
+  interface Window {
+    chrome: {
+      runtime?: any
+      chrome?: boolean
+    }
```
Comment (davidyuk) [inline src/typings/window/index.d.ts:10]:
```suggestion
    } | undefined
```

File: src/utils/aepp-wallet-communication/connection/browser-runtime.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+      description?: string
+      origin?: string
+    }
+    port: Port
+    debug: boolean
+  }) {
+    if (getBrowserAPI().runtime == null) throw new RpcConnectionError('Runtime is not accessible in your environment')
+    this.debug = debug
+    this.connectionInfo = connectionInfo
+    this.port = port ?? getBrowserAPI().runtime.connect(...[connectionInfo.id ?? undefined])
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/connection/browser-runtime.ts:84]:
```suggestion
    this.port = port ?? getBrowserAPI().runtime.connect(connectionInfo.id ?? undefined)
```
won't it work the same?

File: src/utils/aepp-wallet-communication/connection/browser-runtime.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  /**
+   * Disconnect
+   * @instance
+   */
+  disconnect (): void {
+    try {
+      this.port.disconnect()
+    } catch (e) {
+      console.warn('From BrowserRuntimeConnection: ', e)
+    }
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/connection/browser-runtime.ts:96]:
```suggestion
    this.port.disconnect()
```
but this is not related to TS

File: src/utils/aepp-wallet-communication/content-script-bridge.ts [hunk 8]
```diff
@@ -24,63 +24,65 @@
...
+  pageConnection: BrowserWindowMessageConnection
+  extConnection: BrowserRuntimeConnection
+
+  constructor ({ pageConnection, extConnection, allowCrossOrigin = false }: {
+    pageConnection: BrowserWindowMessageConnection
+    extConnection: BrowserRuntimeConnection
+    allowCrossOrigin: boolean }) {
+    if (window == null) throw new UnsupportedPlatformError('Window object not found, you can run bridge only in browser')
+    if (pageConnection == null) throw new MissingParamError('pageConnection required')
+    if (extConnection == null) throw new MissingParamError('extConnection required')
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/content-script-bridge.ts:57]:
```suggestion
```
it should be checked on TS level

File: src/utils/aepp-wallet-communication/rpc/aepp-rpc.ts [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+      return resolveAccountBase(account)
+    }
+
+    if (connection != null) {
+      // Init RPCClient
+      await this.connectToWallet(connection)
+    }
+  },
+  methods: {
+    addresses (): string[] {
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/rpc/aepp-rpc.ts:162]:
```suggestion
    addresses (): Array<EncodedData<'ak'>> {
```

File: src/utils/aepp-wallet-communication/rpc/aepp-rpc.ts [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+      this._ensureAccountAccess()
+      return [this.rpcClient.currentAccount, ...Object.keys(this.rpcClient.accounts.connected)]
+    },
+    /**
+      * Connect to wallet
+      * @function connectToWallet
+      * @instance
+      * @rtype (connection: Object) => void
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/rpc/aepp-rpc.ts:170]:
```suggestion
```

File: src/utils/aepp-wallet-communication/rpc/rpc-client.ts [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+    Object.defineProperties(this, {
+      currentAccount: {
+        enumerable: true,
+        configurable: false,
+        get () {
+          return this.isHasAccounts() === true
+            ? Object.keys(this.accounts.current)[0]
+            : undefined
+        }
+      },
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/rpc/rpc-client.ts:134]: define a TS getter instead

File: src/utils/aepp-wallet-communication/rpc/rpc-client.ts [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+export default class RpcClient {
+  id: string
+  connection: Connection
+  info: RpcClientInfo
+  callbacks: Map<string | number, {resolve: Function, reject: Function}>
+  addressSubscription: string[]
+  accounts: Accounts
+  sendMessage: Function
+  readonly addresses: string[]
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/rpc/rpc-client.ts:89]:
```suggestion
  readonly addresses: Array<EncodedData<'ak'>>
```

File: src/utils/aepp-wallet-communication/rpc/rpc-client.ts [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+   * @param info Info to update (will be merged with current info object)
+   */
+  updateInfo (info: Partial<RpcClientInfo>): void {
+    this.info = { ...this.info, ...info }
+  }
+
+  isHasAccounts (): boolean {
+    return typeof this.accounts === 'object' &&
+      typeof this.accounts.connected === 'object' &&
+      typeof this.accounts.current === 'object'
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/rpc/rpc-client.ts:168]:
```suggestion
      this.accounts.current != null
```

File: src/utils/aepp-wallet-communication/rpc/rpc-client.ts [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+   */
+  hasAccessToAccount (address: string): boolean {
+    return (address.length !== 0) && (this.addresses.find(a => a === address) != null)
+  }
+
+  /**
+   * Check if is connected
+   * @function isConnected
+   * @instance
+   * @rtype () => Boolean
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/rpc/rpc-client.ts:191]:
```suggestion
```

File: src/utils/aepp-wallet-communication/wallet-detector.ts [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * WalletDetector
+ * @alias module:@aeternity/aepp-sdk/es/utils/aepp-wallet-communication/wallet-detector
+ * @param params - Initializer object
+ * @returns WalletDetector
+ */
+export default class WalletDetector {
+  connection: WalletConnection
+  wallets: any
```
Comment (davidyuk) [inline src/utils/aepp-wallet-communication/wallet-detector.ts:38]: It can be at least
```suggestion
  wallets: { [key: string]: any }
```

**Review by davidyuk — CHANGES_REQUESTED** [review]

**Review by subhod-i — APPROVED** [review]

File: src/utils/aepp-wallet-communication/rpc/rpc-client.ts [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+    connection: Connection
+    handlers: [Function, Function]
+    info?: Partial<RpcClientInfo>
+    name?: RpcClientInfo['name']
+    icons?: RpcClientInfo['icons']
+  }) {
+    this.id = id
+    this.connection = connection
+    this.info = { ...info, name, icons }
+
```
Comment (subhod-i) [inline src/utils/aepp-wallet-communication/rpc/rpc-client.ts:117]: extra new line
```suggestion

```

File: src/utils/aepp-wallet-communication/rpc/rpc-client.ts [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+    this.callbacks = new Map()
+    // ['connected', 'current']
+    this.addressSubscription = []
+    this.accounts = {}
+
```
Comment (subhod-i) [inline src/utils/aepp-wallet-communication/rpc/rpc-client.ts:122]: Extra new line
```suggestion
```

File: src/utils/aepp-wallet-communication/rpc/rpc-client.ts [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+    { forceNotification }: { forceNotification?: boolean} = {}): void {
+    if (!isValidAccounts(accounts)) {
+      throw new TypeError('Invalid accounts object. Should be object like: `{ connected: {}, selected: {} }`')
+    }
+    this.accounts = accounts
+    if (!(forceNotification ?? false)) {
+      // Sent notification about account updates
+      this.sendMessage(message(METHODS.updateAddress, this.accounts), true)
+    }
+  }
```
Comment (subhod-i) [inline src/utils/aepp-wallet-communication/rpc/rpc-client.ts:204]:
```suggestion
  setAccounts (accounts: Accounts, forceNotification?: boolean): void {
    if (!isValidAccounts(accounts)) {
      throw new TypeError('Invalid accounts object. Should be object like: `{ connected: {}, selected: {} }`')
    }
    this.accounts = accounts
    if (!(forceNotification ?? false)) {
      // Sent notification about account updates
      this.sendMessage(message(METHODS.updateAddress, this.accounts), true)
    }
  }
```
suggesting to make `forceNotification` an independent param.

File: src/utils/aepp-wallet-communication/wallet-detector.ts [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+ * WalletDetector
+ * @alias module:@aeternity/aepp-sdk/es/utils/aepp-wallet-communication/wallet-detector
+ * @param params - Initializer object
+ * @returns WalletDetector
+ */
+export default class WalletDetector {
+  connection: WalletConnection
+  wallets: Wallets
+
+  constructor ({ connection }: { connection: WalletConnection }) {
```
Comment (subhod-i) [inline src/utils/aepp-wallet-communication/wallet-detector.ts:45]: I don't see why we need to accept a `connection` object here, it can be an independent param.
i see this a lot, Am I missing anything?
```suggestion
  constructor (connection?: WalletConnection) {
```

**Review by subhod-i — CHANGES_REQUESTED** [review]

**Comment by davidyuk** [comment]
closed in favor of #1548
